staging/kos/README.md

```diff
@@ -243,47 +265,57 @@ violating the invariants cannot exist with the easier-to-enforce constraint
+all the subnets and follows the usual Kubernetes controller design pattern
+based on a rate limiting queue and worker goroutines. When a worker processes
+an existing subnet *X*, it does a live list of all the existing subnets
+against the extension API server, and compares *X* against the results of the
+list. If no violation of the invariants is found, *X*'s
+`SubnetStatus.Validated` is set to `true`. The validator also keeps a
+*conflicts cache* which associates each processed subnet *Y*'s namespaced
+name to *Y*'s UID and a list with the namespaced names of the subnets whose
+validation failed because of a conflict with *Y*. If, while validating *X*,
+a conflict with *Y* is found, *Y*'s *UID* is compared to that stored in *Y*'s
+namespaced name *conflicts cache* entry. If they match, *X*'s status is
+updated by adding information on the conflict, and *X*'s namespaced name is
+added to the conflicts cache entry for *Y*'s namespaced name. If they do not
+match, no further action is taken and *X* is processed again after a delay.
+If *Y* is deleted the *conflicts cache* entry for its namespaced name is
+deleted and all the namespaced names in it are enqueued so that their
+subnets can be re-validated: they did not pass validation because of a
+conflict with *Y*, but *Y* no longer exists so the conflict might have
+disappeared and they might be successfully validated now. The same happens
+if a subnet *Y1* with the same namespaced name as *Y* is created shortly
+after *Y*'s deletion and the validator misses the deletion and processes
+*Y1* directly, the only difference being that the *conflicts cache* entry
+for *Y* and *Y1*'s namespaced name is kept and reset by overriding *Y*'s UID
+with *Y1*'s and resetting the list of conflicting subnets' namespaced names.
+Making the addition of *X* to the conflicts cache corresponding to *Y*'s
+namespaced name fail on mistach with the *UID* in that conflict cache
+correctly handles the possible concurrent execution where one goroutine
+processing *X* finds a conflict with *Y*, then *Y* is deleted and immediately
+after another subnet *Y1* with the same namespaced name but different VNI
+and CIDR block is created and a different goroutine processes *Y1*, then the
+first goroutine gets around to trying to add *X* to the conflict cache which
+belonged to *Y* and now belongs to *Y1*; doing that addition in this situation
+would falsely indicate that *X* does not need to be reconsidered until *Y1*
+is deleted.
 
 One apparently odd design choice is having the validator retrieve all the
 subnets with a live list against the API server rather than a cache-based
 list using its Informer. The live list is worse performance-wise, but is
 necessary to avoid race conditions that can lead to two conflicting subnets
-having both `SubnetStatus.Validated` set to true. The subnets validator is a
+having both `SubnetStatus.Validated` set to `true`. The subnets validator is a
```

Comment:
I would s/is a singleton/is intended to be a singleton/

Reply:
Fixed in otherwise unrelated PR #69.